cmd/hook.go

```diff
@@ -460,3 +473,261 @@ func pushOptions() map[string]string {
 	}
 	return opts
 }
+
+func runHookProcReceive(c *cli.Context) error {
+	setup("hooks/proc-receive.log", c.Bool("debug"))
+
+	if len(os.Getenv("SSH_ORIGINAL_COMMAND")) == 0 {
+		if setting.OnlyAllowPushIfGiteaEnvironmentSet {
+			fail(`Rejecting changes as Gitea environment not set.
+If you are pushing over SSH you must push with a key managed by
+Gitea or set your environment appropriately.`, "")
+		} else {
+			return nil
+		}
+	}
+
+	lf, err := os.OpenFile("test.log", os.O_CREATE|os.O_TRUNC|os.O_WRONLY, 0777)
+	if err != nil {
+		fail("Internal Server Error", "open log file failed: %v", err)
+	}
+	defer lf.Close()
+
+	if git.CheckGitVersionAtLeast("2.29") != nil {
+		fail("Internal Server Error", "git not support proc-receive.")
+	}
+
+	reader := bufio.NewReader(os.Stdin)
+	repoUser := os.Getenv(models.EnvRepoUsername)
+	repoName := os.Getenv(models.EnvRepoName)
+	pusherID, _ := strconv.ParseInt(os.Getenv(models.EnvPusherID), 10, 64)
+	pusherName := os.Getenv(models.EnvPusherName)
+
+	// 1. Version and features negotiation.
+	// S: PKT-LINE(version=1\0push-options atomic...)
+	// S: flush-pkt
+	// H: PKT-LINE(version=1\0push-options...)
+	// H: flush-pkt
+
+	rs, err := readPktLine(reader)
+	if err != nil {
+		fail("Internal Server Error", "Pkt-Line format is wrong :%v", err)
+	}
+	if rs.Type != pktLineTypeData {
+		fail("Internal Server Error", "Pkt-Line format is wrong. get %v", rs)
+	}
+
+	const VersionHead string = "version=1"
+
+	if !strings.HasPrefix(rs.Data, VersionHead) {
+		fail("Internal Server Error", "Pkt-Line format is wrong. get %v", rs)
+	}
+
+	hasPushOptions := false
+	response := []byte(VersionHead)
+	if strings.Contains(rs.Data, "push-options") {
```

Comment:
You should split capabilities using space character, and try to match each capability.
If there is a new capability called "push-options-v2", using `string.Contains` is wrong.